### ISSUE TITLE
error while loading shared libraries

### ISSUE THREAD

**Issue by annahutch** [Boyle-Lab/SEMpl#18]
Thank you for developing SEMpl! I am having difficulties running the demo example. I have followed the instructions but get this error message when running the final line:

> ./iterativeSEM -PWM examples/MA0114.1.pwm -merge_file examples/wgEncodeOpenChromDnaseHepg2Pk.narrowPeak -big_wig examples/wgEncodeHaibTfbsHepg2Hnf4asc8987V0416101RawRep1.bigWig -TF_name HNF4A -genome data/hg19 -output results/HNF4A

./iterativeSEM: error while loading shared libraries: libTFMpvalue.so: cannot open shared object file: No such file or directory

---

I have also tried adding SEMpl to my path, e.g. something like 

export PATH=$PATH:/path_to/SEMpl/

but still no luck.

Thanks in advance.
Anna

**Comment by aboyle** [comment]
This error is because the libTFMpvalue.so is missing. Check SEMpl/lib/ to see which so files are there. 

Make sure to follow the build steps exactly (and be sure to use --recurse-submodules when cloning the repo)

Clone a copy of the SEMpl repository and submodules:
`git clone --recurse-submodules https://github.com/Boyle-Lab/SEMpl.git`

Build external libraries:
```
cd SEMpl/lib/libBigWig
make
cd ..
make
mv */*.so .
cd ..
```

**Comment by annahutch** [comment]
Hi, thanks for your response. The libTFMpvalue.so file is in the SEMpl/lib/ directory. I followed these build steps exactly. Thanks, Anna

**Comment by aboyle** [comment]
You should also add the libraries to your LD_LIBRARY_PATH:
export LD_LIBRARY_PATH="/lib/x86_64-linux-gnu:<SEMpl path>/lib";

**Comment by annahutch** [comment]
Hi, 

I am still having difficulty running the program. 

Incase this is of use to future users, these are the extra steps I have had to include so far:

* Add libraries to LD_LIBRARY_PATH:
`export LD_LIBRARY_PATH="/lib/x86_64-linux-gnu:/lib";`

* Create a results directory:
`mkdir results`

* Update G++ compiler:
`module load gcc/9`

I have then downloaded and extracted the bowtie index files :
```
wget https://genome-idx.s3.amazonaws.com/bt/hg19.zip
mv hg19.zip data/
unzip hg19.zip
```

This lists the index files (hg19.1.bt2, hg19.3.bt2, hg19.rev.1.bt2, hg19.2.bt2,  hg19.4.bt2,  hg19.rev.2.bt2) in the directory `SEMpl/data/hg19/`

I then run the following 

`./iterativeSEM -PWM examples/MA0114.1.pwm -merge_file examples/wgEncodeOpenChromDnaseHepg2Pk.narrowPeak -big_wig examples/wgEncodeHaibTfbsHepg2Hnf4asc8987V0416101RawRep1.bigWig -TF_name HNF4A -genome data/hg19/hg19 -output results/HNF4A`

but get the following error message: 
```
Could not locate a Bowtie index corresponding to basename "data/hg19/hg19"
Command: ./bin/bowtie --threads 20 --quiet -a -v 0 -r data/hg19/hg19 ./results/HNF4A/it0/ALIGNMENT/A_pos0.fa 
cache: 0 align: 0
corresponding findMaximumAverageSignalWrapper data is missing!!!!
```

Running the following

`./iterativeSEM -PWM examples/MA0114.1.pwm -merge_file examples/wgEncodeOpenChromDnaseHepg2Pk.narrowPeak -big_wig examples/wgEncodeHaibTfbsHepg2Hnf4asc8987V0416101RawRep1.bigWig -TF_name HNF4A -genome data/hg19/hg19* -output results/HNF4A`

gives the error message: "Could not locate a Bowtie index corresponding to basename "data/hg19/hg19.1.bt2"", but alas this file exists. 

----

I've also tried using alternative bowtie index files:
```
wget ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/h_sapiens_37_asm.ebwt.zip
mv mv h_sapiens_37_asm.ebwt.zip data/
```

which gets me slightly further

`./iterativeSEM -PWM examples/MA0114.1.pwm -merge_file examples/wgEncodeOpenChromDnaseHepg2Pk.narrowPeak -big_wig examples/wgEncodeHaibTfbsHepg2Hnf4asc8987V0416101RawRep1.bigWig -TF_name HNF4A -genome data/h_sapiens_37_asm -output results/HNF4A`

but leaves me with an error after a couple of minutes:

```
Kmer count is 65536
65536
getLength: #AAAACAAAGGCCA# size13
0A to_align: 26379/26379
cache: 0 align: 0
corresponding findMaximumAverageSignalWrapper data is missing!!!!
```

Thanks for your help in advance,
Anna

**Comment by annahutch** [comment]
Hi,

Incase this is of use to future users, these are the steps I had to follow in order to get the example to run.

* Add libraries to LD_LIBRARY_PATH: 
`export LD_LIBRARY_PATH="/lib/x86_64-linux-gnu:/lib";`

* Create a results directory:
`mkdir results`

* Update G++ compiler:
`module load gcc/9`

I then had issues with the Bowtie index files, in that the software was not recognising them. I discovered that this was because I was using Bowtie2 index files, but the software uses an earlier version of Bowtie which does not support Bowtie2 index files. To get around this, I first renamed the existing bowtie in the software and downloaded the latest Bowtie instead:

`wget
https://downloads.sourceforge.net/project/bowtie-bio/bowtie/1.3.0/bowtie-1.3.0-linux-x86_64.zip`

I then unzipped this and created a symlink so that the software would use this version of Bowtie instead:

`ln -s path/to/bowtie-1.3.0-linux-x86_64/bowtie bowtie`

The example now runs correctly.
